### PR TITLE
[yup] Fix type definitions for `default`

### PR DESCRIPTION
**Pull request by iansan5653**
Based on a comment in a previous PR (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44589#issuecomment-637610183), I realized that `default` is typed incorrectly. It is typed as not modifying the schema type, but in reality it adds ~~`null` and~~ `undefined` to the allowable types. This PR makes the types accurate for `default()`:

```ts
// This should have type StringSchema<string | undefined>, but it previously had StringSchema<string>
yup.string().defined().default("abc");

// This should have type StringSchema<string | undefined>, but it previously had StringSchema<string>
yup.string().required().default("abc");
```

Also adds test for this.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. 
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: DEMO: https://codesandbox.io/s/async-feather-w5u7f
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header. (it doesn't)
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by mauricedb** [comment]
Are you sure?  It looks to me that `default(...)` has no effect on what is valid at all.

Both these schemas produce the same output.
```
const schemaWithDefault = boolean().default(false);
console.log("true", schemaWithDefault.isValidSync(true));
console.log("null", schemaWithDefault.isValidSync(null));
console.log("undefined", schemaWithDefault.isValidSync(undefined));

const schemaNoDefault = boolean();
console.log("true", schemaNoDefault.isValidSync(true));
console.log("null", schemaNoDefault.isValidSync(null));
console.log("undefined", schemaNoDefault.isValidSync(undefined));
```

Passing `null` is invalid in both cases.

image

See also: https://github.com/jquense/yup/blob/master/src/mixed.js#L341

**Comment by mauricedb** [comment]
BTW it looks like `defined()` resolves wrong.


```
const schemaDefined = boolean().defined()
console.log("true", schemaDefined.isValidSync(true));
console.log("null", schemaDefined.isValidSync(null));
console.log("undefined", schemaDefined.isValidSync(undefined));
```

it resolves as `const schemaDefined: BooleanSchema<boolean>` but `null` is actually valid. Not what I was expecting from https://github.com/jquense/yup/blob/master/src/mixed.js#L570.

image

**Comment by iansan5653** [comment]
Hmmm....so if you use `validate`/`validateSync` to run validation, you get a different result than if you use `isValid`/`isValidSync`. `default` changes the result of `validate`, but not `isValid`. I guess going by the more correct definition, we should not merge this, at least until we have a way to distinguish in the types between the validation type vs the transformation type.

Interesting observation in your second comment; I can reproduce. I believe that's a bug in Yup - calling `defined` shouldn't automatically make a field nullable.

**Comment by mauricedb** [comment]
I am seeing the same behavior with `isValidSync()` and `validateSync()`. The `null` fails validation and `undefined` and `true` pass,

```
const schemaWithDefault = boolean().default(false);
console.log("true", schemaWithDefault.isValidSync(true));
console.log("true passes", schemaWithDefault.validateSync(true));
console.log("null", schemaWithDefault.isValidSync(null));
try {
  console.log("null passes", schemaWithDefault.validateSync(null));
} catch (error) {
  console.log("null throws");
}
console.log("undefined", schemaWithDefault.isValidSync(undefined));
console.log("undefined passes", schemaWithDefault.validateSync(undefined));

const schemaNoDefault = boolean();
console.log("true", schemaNoDefault.isValidSync(true));
console.log("true passes", schemaNoDefault.validateSync(true));
console.log("null", schemaNoDefault.isValidSync(null));
try {
  console.log("null passes", schemaNoDefault.validateSync(null));
} catch (error) {
  console.log("null throws");
}
console.log("undefined", schemaNoDefault.isValidSync(undefined));
console.log("undefined passes", schemaNoDefault.validateSync(undefined));
```

image

**Comment by iansan5653** [comment]
OK so after playing with it I think that it does not add `null`, so that part of my PR is incorrect. However it *does* add `undefined`, which you can see if you chain `default` after `required` or `defined`:

```ts
console.log(" --- without required or defined --- ")

const schema = number();
console.log("number(): undefined", schema.isValidSync(undefined));
console.log("number(): null", schema.isValidSync(null));

const schemaWithDefault = number().default(7);
console.log("number().default(7): undefined", schemaWithDefault.isValidSync(undefined));
console.log("number().default(7): null", schemaWithDefault.isValidSync(null));

console.log(" --- with defined --- ")

const schemaDefined = number().defined();
console.log("number().defined(): undefined", schemaDefined.isValidSync(undefined));
console.log("number().defined(): null (bug)", schemaDefined.isValidSync(null));

const schemaDefinedWithDefault = number().defined().default(7);
console.log("number().defined().default(7): undefined", schemaDefinedWithDefault.isValidSync(undefined));
console.log("number().defined().default(7): null", schemaDefinedWithDefault.isValidSync(null));

console.log(" --- with required --- ")

const schemaRequired = number().required();
console.log("number().required(): undefined", schemaRequired.isValidSync(undefined));
console.log("number().required(): null (bug)", schemaRequired.isValidSync(null));

const schemaRequiredWithDefault = number().required().default(7);
console.log("number().required().default(7): undefined", schemaRequiredWithDefault.isValidSync(undefined));
console.log("number().required().default(7): null", schemaRequiredWithDefault.isValidSync(null));
```

image

So it actually looks like the change is valid but I need to remove `null` from the modified types.

**Comment by mauricedb** [comment]
It isn't so mutch that it adds `undefined` as a valid value but when the value being validated is `undefined` the `default()` value is being used instead. So it depends on what you pass in there an dif that is valid. Use `.default(undefined)` or `.default(true)` and validation fails again.

I was also expecting `.strict(true)` to have an inpact here but it doesn't with `undefined`. To my suprise it does inpact validating `null`.

```
console.log(" --- with default 7 --- ");
const schemaDefinedWithDefault = number()
  .defined()
  .default(7);
console.log(
  "number().defined().default(7): undefined",
  schemaDefinedWithDefault.isValidSync(undefined)
);
console.log(
  "number().defined().default(7): null",
  schemaDefinedWithDefault.isValidSync(null)
);

console.log(" --- with default undefined --- ");
const schemaDefinedWithDefaultUndefined = number()
  .defined()
  .default(undefined);
console.log(
  "number().defined().default(undefined): undefined",
  schemaDefinedWithDefaultUndefined.isValidSync(undefined)
);
console.log(
  "number().defined().default(undefined): null",
  schemaDefinedWithDefaultUndefined.isValidSync(null)
);

console.log(" --- with default 7 and strict --- ");
const schemaWithDefaultStrict = number()
  .default(7)
  .strict(true);
console.log(
  "number().default(7).strict(true): undefined",
  schemaWithDefaultStrict.isValidSync(undefined)
);
console.log(
  "==> number().default(7).strict(true): null",
  schemaWithDefaultStrict.isValidSync(null)
);
```

image

**Comment by iansan5653** [comment]
> It isn't so mutch that it adds undefined as a valid value but when the value being validated is undefined the default() value is being used instead.

You're not wrong, but as long as we constrain the passed `default` value to `extends T` (`T` being the schema type), it does essentially add `undefined` as an allowed value since `undefined` will always pass validation.

With regards to `strict`, I don't think it's impacting `null` at all. That's just another manifestation of the bug with `defined` allowing `null`.

In this example you can see `null` is invalid unless you include `defined`, then it's valid no matter what:
```ts
import { number } from "yup";

console.log(
  "number().defined().default(7): null",
  number()
    .defined()
    .default(7)
    .isValidSync(null)
);
console.log(
  "number().defined().default(7).strict(): null",
  number()
    .defined()
    .default(7)
    .strict()
    .isValidSync(null)
);
console.log(
  "number().default(7): null",
  number()
    .default(7)
    .isValidSync(null)
);
console.log(
  "number().default(7).strict(): null",
  number()
    .default(7)
    .strict()
    .isValidSync(null)
);
```
image

**Comment by iansan5653** [comment]
Lol, sorry for all the back and forth on this. I thought about it for a long time and realized I'm misunderstanding the way the types should work. Schemas work by taking `any` and modifying it with a series of transformations *before* validating it. The schema type is not at all related to the transformations, only to the ultimate validated type.

Therefore, while `default` may ensure that the post-transforms value is never `undefined`, it shouldn't be seen as excluding `undefined` from the schema type, because it's not *invalidating* `undefined`. If you want to invalidate `undefined`, you just use `.defined()`.

So this is totally OK and should be how things are done:

```ts
yup.string().default("example").defined();
```

For any future reader who may not understand this (it took me admittedly way too long) it's easier if you look at custom transformation functions:

```
// This makes sure the value is never "a"
const customTransform = value => value === "a" ? undefined : value;

// With this schema, we know the resulting value will never be "a", but that doesn't mean
// we made "a" an invalid value!
const schema = string().transform(customTransform);

// This is weird but technically totally fine
// (assuming the default is applied after the transform, but that's not relevant for the example)
const schema = string().transform(customTransform).default("a");
```

So in short this PR really isn't valid.
